pallets/nomad/da-bridge/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+
+	// Genesis config
+	#[pallet::genesis_config]
+	pub struct GenesisConfig {}
+
+	#[cfg(feature = "std")]
+	impl Default for GenesisConfig {
+		fn default() -> Self { Self {} }
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig {
+		fn build(&self) {}
+	}
+
+	// Hooks
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_finalize(block_number: T::BlockNumber) {
+			let last = block_number.checked_sub(&1u32.into()).unwrap();
+
+			// Store every finalized block's corresponding hash
+			let hash = frame_system::Pallet::<T>::block_hash(last);
+			FinalizedBlockNumberToBlockHash::<T>::insert(last, hash);
+		}
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		DataRootDispatched {
+			destination_domain: u32,
+			recipient_address: H256,
+			block_number: T::BlockNumber,
+			data_root: H256,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		InitializationError,
+		BlockNotFinal,
+		HashOfBlockNotMatchBlockNumber,
+		DABridgeMessageExceedsMaxMessageSize,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		[u8; 32]: From<T::AccountId>,
+		H256: From<T::Hash>,
+		u32: From<T::BlockNumber>,
+	{
+		/// Dispatch a data root message to the home if the header is valid.
+		#[pallet::weight(100)]
+		pub fn try_dispatch_data_root(
+			origin: OriginFor<T>,
+			destination_domain: u32,
+			recipient_address: H256,
+			header: T::Header,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			Self::ensure_valid_header(&header)?;
+			Self::do_dispatch_data_root(destination_domain, recipient_address, header)
+		}
+	}
+
+	impl<T: Config> Pallet<T>
+	where
+		[u8; 32]: From<T::AccountId>,
+		H256: From<T::Hash>,
+		u32: From<T::BlockNumber>,
+	{
+		/// Dispatch a data root message for a valid header.
+		fn do_dispatch_data_root(
+			destination_domain: u32,
+			recipient_address: H256,
+			header: T::Header,
+		) -> DispatchResult {
+			let mut block_number = *header.number();
+			let data_root = header.data_root();
+
+			let message: DABridgeMessages = DataRootMessage {
+				block_number: block_number.into(),
+				data_root,
+			}
+			.into();
+
+			let body: BoundedVec<u8, T::MaxMessageBodyBytes> = message
+				.encode()
+				.try_into()
+				.map_err(|_| Error::<T>::DABridgeMessageExceedsMaxMessageSize)?;
+
+			Home::<T>::do_dispatch(
+				T::DABridgePalletId::get(),
+				destination_domain,
+				recipient_address,
+				body,
+			)?;
+
+			// Clear previous block_number to hash mappings, starting at the
+			// current submitted block_number going backwards. Because the
+			// runtime maps _every_ block number's hash, the sequence is always
+			// contiguous. If we are tracing backwards and find a block number
+			// that doesn't have a hash, we know everything before it has been
+			// cleared.
+			while FinalizedBlockNumberToBlockHash::<T>::contains_key(block_number) {
```

Comment:
:bangbang: **Security** : Unbounded operation. It is impossible to determine the cost of `try_dispatch_data_root`.

Reply:
How would you suggest clearing the mapping then in bounded way? We run into the same problem in the `home` pallet when clearing the root <> index mappings

Reply:
Would you prefer to just have loop have bounded number of iterations? And how to parametrize that?